lang/driver/src/database.rs

```diff
@@ -186,57 +201,103 @@ impl Database {
+
+        // Compute the type info table
+        let mut info_table = TypeInfoTable::default();
+        let mod_info_table = self.type_info_table(uri)?;
+        info_table.insert(uri.clone(), mod_info_table);
+        for dep_url in deps {
+            let mod_info_table = self.type_info_table(&dep_url)?;
+            info_table.insert(dep_url.clone(), mod_info_table);
+        }
 
-        self.load_ast(uri)
+        // Typecheck module
+        let ust = self.ust(uri).map(|x| (*x).clone())?;
+        let ast = elaborator::typechecker::check_with_lookup_table(Rc::new(ust), &info_table)
+            .map(Arc::new)
+            .map_err(Error::Type);
+        self.ast.insert(uri.clone(), ast.clone());
+        ast
     }
 
-    pub fn load_ast(&mut self, uri: &Url) -> Result<Arc<ast::Module>, Error> {
-        log::trace!("Loading AST: {}", uri);
+    // Core API: info_by_id
+    //
+    //
 
-        match self.ast.get_unless_stale(uri) {
-            Some(ast) => ast.clone(),
-            None => {
-                log::trace!("AST is stale, reloading");
-                let ust = self.ust(uri).map(|x| (*x).clone())?;
-
-                // Compute the dependencies
-                let empty_vec = Vec::new();
-                let direct_dependencies = self.deps.get(uri).unwrap_or(&empty_vec).clone();
-
-                // Compute the type info table
-                let mut info_table = TypeInfoTable::default();
-                let mod_info_table = self.type_info_table(uri)?;
-                info_table.insert(uri.clone(), mod_info_table);
-                for dep_url in direct_dependencies {
-                    let mod_info_table = self.type_info_table(&dep_url)?;
-                    info_table.insert(dep_url.clone(), mod_info_table);
-                }
-                let ast =
-                    elaborator::typechecker::check_with_lookup_table(Rc::new(ust), &info_table)
-                        .map(Arc::new)
-                        .map_err(Error::Type);
-
-                self.ast.insert(uri.clone(), ast.clone());
-                if let Ok(module) = &ast {
-                    let (info_lapper, item_lapper) = collect_info(module.clone());
-                    self.info_by_id.insert(uri.clone(), info_lapper);
-                    self.item_by_id.insert(uri.clone(), item_lapper);
-                }
-                ast
+    pub fn info_by_id(&mut self, uri: &Url) -> Result<Lapper<u32, Info>, Error> {
+        match self.info_by_id.get_unless_stale(uri) {
+            Some(infos) => {
+                log::debug!("Found info_by_id in cache: {}", uri);
+                Ok(infos.clone())
             }
+            None => self.recompute_info_by_id(uri),
         }
     }
 
+    fn recompute_info_by_id(&mut self, uri: &Url) -> Result<Lapper<u32, Info>, Error> {
+        log::debug!("Recomputing info_by_id for: {}", uri);
+        let ast = self.ast(uri)?;
+        let (info_lapper, _item_lapper) = collect_info(ast.clone());
```

Comment:
I wonder if `collect_info` should also be split, then.

Reply:
I have to rewrite the logic for collecting info anyway, since it currently doesn't work for jump-to-definition to imported modules. I can take a look when fixing that issue.